### ISSUE TITLE
Support ignoring gitignore files in ls output

### ISSUE THREAD

**Issue by yerlaser** [nushell/nushell#8537]
### Related problem

I often need to do recursive file searches using `ls **/*` and `glob **/*`
Unfortunately, my workspaces contain `target` and `node_modules` folders with many small files and that greatly slows down the searches.
Almost always I want to ignore such folders and usually (but not always) they are already listed in `.gitignore` file.

### Describe the solution you'd like

Allow `ls` and `glob` commands to accept a parameter `--ignore (-i)` that would ignore files and folders specified in corresponding `.gitignore` and `.ignore/.fdignore` files like how `ripgrep` and `fd` commands work.

### Additional context and details

References:
https://github.com/BurntSushi/ripgrep
https://github.com/sharkdp/fd

**Comment by dmatos2012** [comment]
Ill look into this, if thats ok @fdncred :). Seeing what are our options, and whether we can suppor this at least in the `glob` command!

**Comment by yerlaser** [comment]
Thank you!

**Comment by fdncred** [comment]
sounds good @dmatos2012 - good luck.

**Comment by bobhy** [comment]
Maybe we don't need a new feature here?

`> glob ** --not [target/**, .git/**, tests/**]`  seems to be well behaved.  If you've got a standard set of excluded paths, you could put them in an environment variable...

**Comment by yerlaser** [comment]
Yes, I use such exclusions. But, the list becomes very large and unmanageable. For instance, on Mac you have .DS_Store files, in Python projects .pyc etc. etc.
Therefore often I prefer to use `fd`, an excellent utility that ignores those files by default.
But, having it in `ls` would greatly help because I like the tabular data that I can further filter as needed.

**Comment by yerlaser** [comment]
> I was sort of working on it a long time ago, but dropped it as I realized that the easiest was to add the crate [ignore](https://crates.io/crates/ignore), and then had an MVP working. Butt thats an extra dependency, and I know `nu` now wants to reduce the number of dependencies, so didnt look further into it sorry :(

Sad, but understandable.
Thanks for trying though.

**Comment by fdncred** [comment]
I was thinking the same thing about using the `ignore` crate but ignoring files would need to be configurable. I'd probably accept a PR to make `ls` better, but "better" is in the eye of the beholder. Other maintainers probably have different opinions.

IMO, I think `ls` should move off of using `nu-glob` and use the [jwalk](https://github.com/Byron/jwalk) crate since it can (1) stream data (2) sort while streaming (3) uses rayon threads for speed (4) uses the `ignore` crate.  I've attempted this a few times and it was harder than I wanted it to be. I'd be jazzed if someone took on the big task of making `ls` work this way.

**Comment by bobhy** [comment]
Let's back up a bit and brainstorm what kind of file enumeration tool we'd like to have before we dive too quickly into the tactics of building it.  See also similar idea being proposed (and worked on) in #10650 by @skyvine.  

I don't think it's wise to invest more in `ls` specifically for this use case (even though it's near and dear to all developers).  There will inevitably be another task-specific scenario that's desired later.  Nushell is a great environment for integrating specialized tools into custom solutions, let's focus on enhancing that instead!

Our file processing tools should be adapted to accept the enumeration performed by some external tool.  See https://github.com/nushell/nushell/issues/9558#issuecomment-1755199503 for a proposal to allow `cp`, `mv`, `rm` to accept their *source* list from pipeline input rather than exclusively on the command line.  That sounds promising.

For fancy file enumeration,  [`fd`](https://github.com/sharkdp/fd) does look nice (thanks for the tip, @yerlaser).  How could it be better integrated with nushell?  I'm thinking if its outputs could easily be converted into a stream of nushell table entries?  Hmm, what if you could define an output type in `extern` command?  That would integrate outputs from  *any* external command.  I'll take a look at that!

It sounds like the combination of all these nascent efforts would make a significant enhancement to the nushell environment!  What say you?

**Comment by bobhy** [comment]
Let's get a bit clearer on the ask, to help undestand how appealing an idea it is:
You are asking whether `ls` could have an option to ignore all files matching `.gitignore`, right?
Would you use it if `glob` had that option?
And are there any other favorite capabilities of `fd` that you'd like to see in either of these commands?
Do you use its regex patterns?  Do you (ever) use the multiple globs in parallel capability?
Have you ever tried making `node_modules` (or `target`) a hidden file (e.g `.target`) to ignore them by default?

**Comment by yerlaser** [comment]
Personally I would love it if `ls` had this functionality.
I found out that that 99% of my usage of `fd` is exactly due to this feature.
I rarely use other features of `fd` and very rarely search for regex patterns
Even if I do, it's because `ls` doesn't filter out _ignore_ files to begin with.
I once even had a script that ran `ls` on each line that `fd` returns to get a list with file metadata, flatten it and then continue processing the stream further in `nu`.

Regarding hiding temp folders, yes, sure I could do it. And also putting global exclusions into `glob` is an option, but they all mean yet more stuff to maintain. That's why I tend to use `fd`, because it _just works_.
Would be perfect if `ls` did the same.

However, as I said earlier, it's my personal use-case and I don't want to demand this if the core developers are against.
Nushell has many other wonderful features that compensate for the lack of this one.

**Comment by yerlaser** [comment]
To give you another perspective, VSCode (which is, of course, developer oriented, but still relevant) hides _ignore_ files both from the tree view in the sidebar and from the searches by default. So, they also understand the benefits of uncluttered view.

I also often type simple `fd` without a pattern in a folder to have a quick overview of the files that I care without the clutter. That's why I wish `ls` could do it.
In those very rare cases when I do care about _ignore_ files, usually because I want to clean up, I just run `ls -a`.

**Comment by Ecorous** [comment]
This can (rather) easily be done using `glob --exclude (open .gitignore | lines | where not ($it starts-with "#"))` or similar in a custom command. Don't think any of this is relevant anymore, can reopen if it is

**Comment by hkalbasi** [comment]
I think it still worth a command. Handling `.gitignore` files is not simple. There is nested `.gitignore` files in subdirectories, `!not/ignore` rules, and many other details. Handling all of these edge cases is time-consuming, error-prone, and slow in nushell. So I would like to have it as a flag on ls or another custom command.

**Comment by fdncred** [comment]
it's not going to happen with `ls`. glob has --exclude as already pointed out. i'm pretty sure the `idx *` commands using the `ignore` crate so there's options with the idx commands.

**Comment by hkalbasi** [comment]
`idx files .` is exactly what I wanted, thanks! Sorry for offtopic question, is it possible to block on `idx init .`? I currently added a `sleep 50ms` to my script to handle it.

**Comment by fdncred** [comment]
`help idx init`
